llms/mlx_lm/server.py

```diff
@@ -0,0 +1,284 @@
+                    max_tokens,
+                    top_p=top_p,
+                )
+            )
+            text = _tokenizer.decode(tokens)
+            response = {
+                "id": chat_id,
+                "object": "chat.completion",
+                "created": int(time.time()),
+                "model": requested_model,
+                "system_fingerprint": f"fp_{uuid.uuid4()}",
+                "choices": [
+                    {
+                        "index": 0,
+                        "message": {
+                            "role": "assistant",
+                            "content": text,
+                        },
+                        "logprobs": None,
+                        "finish_reason": None,
+                    }
+                ],
+                "usage": {
+                    "prompt_tokens": len(prompt),
+                    "completion_tokens": len(tokens),
+                    "total_tokens": len(prompt) + len(tokens),
+                },
+            }
+
+            return response
+        else:
+            self.send_response(200)
+            self.send_header("Content-type", "text/event-stream")
+            self.send_header("Cache-Control", "no-cache")
+            self.end_headers()
+            accumulated_tokens = []
+            generated_text = ""
+            for token in generate(
+                prompt,
+                _model,
+                temperature,
+                stop_id_sequences,
+                eos_token_id,
+                max_tokens,
+                top_p=top_p,
+            ):
+                # This is a workaround because the llama tokenizer omitted spaces during decoding token by token.
+                accumulated_tokens.append(token)
+                cur = _tokenizer.decode(accumulated_tokens)
+                next_chunk = cur[len(generated_text) :]
+                generated_text = cur
+                response = {
+                    "id": chat_id,
+                    "object": "chat.completion.chunk",
+                    "created": int(time.time()),
+                    "model": requested_model,
+                    "system_fingerprint": f"fp_{uuid.uuid4()}",
+                    "choices": [
+                        {
+                            "index": 0,
+                            "delta": {"role": "assistant", "content": next_chunk},
+                            "logprobs": None,
+                            "finish_reason": None,
+                        }
+                    ],
+                }
+                try:
+                    self.wfile.write(f"data: {json.dumps(response)}\n\n".encode())
+                    self.wfile.flush()
+                except Exception as e:
+                    print(e)
+                    break
+
+            self.wfile.write(f"data: [DONE]\n\n".encode())
+            self.wfile.flush()
+
+
+def run(server_class=HTTPServer, handler_class=APIHandler):
+    host = os.getenv("HOST", "127.0.0.1")
```

Comment:
建议将 `HOST` 重命名为 `MLX_SERVER_HOST`，当系统的环境变量中存在大量的定义时，明确的命名可以很方面的知道这个环境变量的用途。同样的，`PORT` 也建议命名为 `MLX_SERVER_PORT` .

Reply:
Yeah, I use the host and port for simplicity since it's mainly for development and testing purposes. However, there is a chance that the environment variable may conflict with other apps running on the development machine. I will update it to specify the host and port via command args.